### PR TITLE
Feature/576 As a guardian i would like to be able to renme choiceboards

### LINKED ISSUE #576: As a Guardian i would like to be able to rename choiceboards.
**Describe the solution you'd like**
Guardians should be able to rename specific choiceboards of citizens.

### PR DESCRIPTION
**Pull request by ggAlexS**
Renaming choiceboards have been implemented 

fixes #576

### CODE REVIEW

**Comment by ggAlexS** [comment]
needs the api-client to be updated for the linter to accept the new variable

**Comment by Determinatoro** [comment]
Please use the following checklist, @AdrianPlesner. Feel free to check it off as you go along.

**Code Design**

- [x] The code is in the right place? (Both in terms of folder structure and class structure)
- [x] The code is not over-engineered. Examples of over-engineering: Implemented behavior for future problems
- [x] This code could not, to the best of my knowledge, have reused existing code
- [x] The code does not introduce functionality that is unnecessary for solving the problem

**Code Readability**

- [x] Names are meaningful and self-documenting
- [x] It is understandable, by reading the code, what it does
- [x] The code is simple and readable. i.e. it contains no "hack", or obscure solution
- [x] Error messages in the code are understandable

**Code Maintainability**

- [x] Has all the added code been properly commented

Only check one of the items below:

- [ ] The functionality does **not** need be covered by tests
- [x] The functionality needs to be covered by tests

Only use if the functionality needs to be covered by tests:
- [ ] All functionality is covered by tests
- [x] The tests has sensible names
- [x] By reading the tests you know what they cover
- [ ] The tests cover sensible cases. Both happy paths and exception paths
- [ ] The tests cover the full functionality. i.e., The tests fails if some of the requested functionality is missing

**Functionality**

For this part, open the app, and test:

- [x] The features that the code claims to implement, are actually implemented
- [x] The code does not look like it has bugs
- [x] All the new screens are reachable through in-app navigation

**Comment by Determinatoro** [comment]
Please use the following checklist, @social2dk. Feel free to check it off as you go along.

**Code Design**

- [x] The code is in the right place? (Both in terms of folder structure and class structure)
- [x] The code is not over-engineered. Examples of over-engineering: Implemented behavior for future problems
- [x] This code could not, to the best of my knowledge, have reused existing code
- [x] The code does not introduce functionality that is unnecessary for solving the problem

**Code Readability**

- [x] Names are meaningful and self-documenting
- [x] It is understandable, by reading the code, what it does
- [x] The code is simple and readable. i.e. it contains no "hack", or obscure solution
- [x] Error messages in the code are understandable

**Code Maintainability**

- [x] Has all the added code been properly commented

Only check one of the items below:

- [ ] The functionality does **not** need be covered by tests
- [x] The functionality needs to be covered by tests

Only use if the functionality needs to be covered by tests:
- [x] All functionality is covered by tests
- [x] The tests has sensible names
- [x] By reading the tests you know what they cover
- [x] The tests cover sensible cases. Both happy paths and exception paths
- [x] The tests cover the full functionality. i.e., The tests fails if some of the requested functionality is missing

**Functionality**

For this part, open the app, and test:

- [x] The features that the code claims to implement, are actually implemented
- [x] The code does not look like it has bugs
- [x] All the new screens are reachable through in-app navigation

**Comment by ZohraAmini** [comment]
please merge develop into this branch, so we can test.

**Review by Huften — COMMENTED** [review]
Is it working as intended for you guys?
Because for me, when I add a "valgmulighed" to the choice board, and return to weekplanner screen, the choiceboard activity is not correctly displayed. When I then return to the same activity, the choice board title is now deleted.

**Comment by ggAlexS** [comment]
> 
> 
> Is it working as intended for you guys?
> Because for me, when I add a "valgmulighed" to the choice board, and return to weekplanner screen, the choiceboard activity is not correctly displayed. When I then return to the same activity, the choice board title is now deleted.

works for me

**Comment by AdrianPlesner** [comment]
This solution does not fit the provided prototype for the issue:

**Comment by ggAlexS** [comment]
> This solution does not fit the provided prototype for the issue:

The design in the prototype, is not optimised for use on mobile units with touch. Due to fact that it is to big, and the on screen keyboard totally squashes it

**Comment by AdrianPlesner** [comment]
> The design in the prototype, is not optimised for use on mobile units with touch. Due to fact that it is to big, and the on screen keyboard totally squashes it

Has this been approved by PO?
It should be possible to make a solution where the keyboard is added as an extension to the screen instead ike other screens, see example. Maybe make the screen scrollable?
Simulator Screen Shot - iPad Air (4th generation) - 2020-10-30 at 09 41 56

**Comment by JeppeKLau** [comment]
> 
> 
> > The design in the prototype, is not optimised for use on mobile units with touch. Due to fact that it is to big, and the on screen keyboard totally squashes it
> 
> Has this been approved by PO?
> It should be possible to make a solution where the keyboard is added as an extension to the screen instead ike other screens, see example. Maybe make the screen scrollable?

First a quick question, why are you even interested in this pull request? You aren't a reviewer or a developer, so your interest makes no sense to me. Second, we attempted to implement this but due to the stupid way the activity screen has been implemented it was not possible without a full refactor of the screen, which would be way out of the scope for the issue.

**Comment by AdrianPlesner** [comment]
> First a quick question, why are you even interested in this pull request? 

I'm in the same group as reviewer @Huften who asked me to help him review this due to personal reasons.

**Comment by JeppeKLau** [comment]
> 
> 
> > First a quick question, why are you even interested in this pull request?
> 
> I'm in the same group as reviewer @Huften who asked me to help him review this due to personal reasons.

Has this been cleared with the process group?

**Comment by AdrianPlesner** [comment]
> > I'm in the same group as reviewer @Huften who asked me to help him review this due to personal reasons.
> 
> Has this been cleared with the process group?

Not yet, and this shouldn't really be necessary. We are just trying to give a review as soon as possible. 😃

File: lib/screens/show_activity_screen.dart [hunk 1]
```diff
@@ -332,6 +325,20 @@ class ShowActivityScreen extends StatelessWidget {
     return Card(
         child: Column(children: <Widget>[
       const Center(child: Padding(padding: EdgeInsets.all(8.0))),
+          Visibility(
+              visible: _activity.isChoiceBoard,
+              child: TextFormField(
+                key: const Key('ChoiceBoardNameText'),
+                initialValue: _activity.choiceBoardName,
+                textAlign: TextAlign.center,
+                onFieldSubmitted: (String text)
```
Comment (AdrianPlesner) [inline lib/screens/show_activity_screen.dart]: This field is never submitted, meaning that it is never actually saved, leaving to with a null-value. You should add a submit button, like suggested in the prototype, to actually save the given title. Further, this should either make sure that a title is given, or insert a default value if not.

**Comment by JeppeKLau** [comment]
> 
> Has this been approved by PO?
> It should be possible to make a solution where the keyboard is added as an extension to the screen instead ike other screens, see example. Maybe make the screen scrollable?
> 

The design has been approved by PO and they recognized the need for a refactor of the activity screen.

>
>This field is never submitted, meaning that it is never actually saved
>

The field is submitted when the user hits the 'Enter' key on the keyboard. 

>
>You should add a submit button
>

You are right, this should be added and it will.

>
>leaving to with a null-value.
>

No null-values are assigned to the name of the choice board, as the web api gives the string "ChoiceBoard" as the base value for a choice board if it has no name. And remember that the empty string, aka "", and null are not the same.

**Comment by AdrianPlesner** [comment]
>No null-values are assigned to the name of the choice board, as the web api gives the string "ChoiceBoard" as the base value for a choice board if it has no name

Have you made changes to the api to make this work?

**Comment by JeppeKLau** [comment]
> 
> 
> > No null-values are assigned to the name of the choice board, as the web api gives the string "ChoiceBoard" as the base value for a choice board if it has no name
> 
> Have you made changes to the api to make this work?

Yes of course. This feature has been implemented all the way from front to back from the ground up as it wasn't in the original functionality.

**Comment by AdrianPlesner** [comment]
> Yes of course. This feature has been implemented all the way from front to back from the ground up as it wasn't in the original functionality.

Okay then, this was the reason we couldn't get your solution to work properly. With the right version of the API it seems to be working fine. 
>
>>You should add a submit button
>
>You are right, this should be added and it will.
> 

Are you working on this within this pull request?

**Review by social2dk — DISMISSED** [review]
Remember to also get the matching web-api thing, or this does not work

**Comment by LivHolm** [comment]
@ggAlexS Du eller en i din gruppe er ansvarlig for at resolve de merge conflikter der er.
Der mangler også et review fra @AdrianPlesner.

**Comment by ggAlexS** [comment]
> @ggAlexS Du eller en i din gruppe er ansvarlig for at resolve de merge conflikter der er.
> 
> Der mangler også et review fra @AdrianPlesner.

Tror det her er lagt lidt på is af PO

**Review by AdrianPlesner — CHANGES_REQUESTED** [review]
A save and revert button should be added, as suggested by the prototype, which also adds a way to submit that is not the enter button, which is illogical in apps.

**Comment by puchasa** [comment]
> > @ggAlexS Du eller en i din gruppe er ansvarlig for at resolve de merge conflikter der er.
> > Der mangler også et review fra @AdrianPlesner.
> 
> Tror det her er lagt lidt på is af PO

Det mener PO ikke at det er.

> A save and revert button should be added, as suggested by the prototype, which also adds a way to submit that is not the enter button, which is illogical in apps.

Der er fejl i prototypen da PO har glemt at tage højde for et pop-up tastatur. Derfor ignorerer vi i dette tilfælde at prototypen ikke er blevet fulgt og fokuserer i stedet på at få ting til at virke, således at næste semester kan nøjes med at fokusere på designet. PO vil lave en ny prototype og oprette et issue med det nye design som kan blive inkuderet i handover.

**Comment by AdrianPlesner** [comment]
Knappen behøver måske heller ikke se ud præcist som i prototypen, men derfor burde der stadig være en submit knap så man ikke skal gøre det igennem tasteturet. Desuden er fortryd/hent standard funktionalitet vel noget der skal til for at det "virker"

**Comment by ggAlexS** [comment]
Merge conflicts er løst, så den er klar til merge

**Comment by Kaliahh** [comment]
PO er enige med Adrian om at der i det mindste skal være en knap til at submitte. Vi fixer designet i et senere issue, men det kunne bare være rart med noget der virker.

**Comment by puchasa** [comment]
> PO er enige med Adrian om at der i det mindste skal være en knap til at submitte. Vi fixer designet i et senere issue, men det kunne bare være rart med noget der virker.

@ggAlexS vil du lige kigge på det?

**Comment by ggAlexS** [comment]
> 
> 
> PO er enige med Adrian om at der i det mindste skal være en knap til at submitte. Vi fixer designet i et senere issue, men det kunne bare være rart med noget der virker.



> 
> 
> > PO er enige med Adrian om at der i det mindste skal være en knap til at submitte. Vi fixer designet i et senere issue, men det kunne bare være rart med noget der virker.
> 
> @ggAlexS vil du lige kigge på det?

værsgo

**Review by AdrianPlesner — CHANGES_REQUESTED** [review]
Det lader til at virke nu, jeg kunne dog godt tænke mig som minimun at det bliver testet er der på weekplan_screen kan findes et choiceboard med et andet navn end "Choiceboard", og også gerne at det viste navn på weekplan_screen faktisk bliver ændret når man givet et choiceboard et nyt navn.

De eneste tests der er indtil videre tester kun at der er et tekstfelt når det er et choiceboard, og at det er muligt at indsætte tekst i det.

**Comment by AdrianPlesner** [comment]
Desuden har jeg også observeret at når man tilføjer/slettet/kopierer aktiviteter i weekplanen, bliver alle choiceboard navne reset til "Choiceboard". Dette skyldes navnet ikke bliver taget med under aktiviteter kopieres under UpdateWeek i api'et. Synderen er en constructor i GirafRest/Shared/SharedMethods.cs linje 102. Api delen af det burde nok have været mere grundigt testet, så var denne fejl blevet fanget.

**Comment by ggAlexS** [comment]
> 
> 
> Desuden har jeg også observeret at når man tilføjer/slettet/kopierer aktiviteter i weekplanen, bliver alle choiceboard navne reset til "Choiceboard". Dette skyldes navnet ikke bliver taget med under aktiviteter kopieres under UpdateWeek i api'et. Synderen er en constructor i GirafRest/Shared/SharedMethods.cs linje 102. Api delen af det burde nok have været mere grundigt testet, så var denne fejl blevet fanget.

der bør nok laves et nyt issue på web-api'en til det her

File: test/screens/show_activity_screen_test.dart [hunk 2]
```diff
@@ -875,6 +875,29 @@ void main() {
...
+  });
+
+  testWidgets('ChoiceBoard name can be changed', (WidgetTester tester)
+  async {
+    authBloc.setMode(WeekplanMode.guardian);
+    await tester.pumpWidget(MaterialApp(home:
+    ShowActivityScreen(mockActivity, mockUser)));
+    await tester.pump();
+    await tester.enterText(
+        find.byKey(const Key('ChoiceBoardNameText')), 'nametest');
```
Comment (AdrianPlesner) [inline test/screens/show_activity_screen_test.dart:949]: Hvis jeg ikke tager helt fejl, så er titlen på det her choiceboard "nametest" i forvejen, så det at sætte den til "nametest" igen bekræfter ikke rigtigt at det virker at ændre den?

**Review by AdrianPlesner — CHANGES_REQUESTED** [review]
Testen virker lidt underligt for mig.

**Comment by puchasa** [comment]
> > Desuden har jeg også observeret at når man tilføjer/slettet/kopierer aktiviteter i weekplanen, bliver alle choiceboard navne reset til "Choiceboard". Dette skyldes navnet ikke bliver taget med under aktiviteter kopieres under UpdateWeek i api'et. Synderen er en constructor i GirafRest/Shared/SharedMethods.cs linje 102. Api delen af det burde nok have været mere grundigt testet, så var denne fejl blevet fanget.
> 
> der bør nok laves et nyt issue på web-api'en til det her

@ggAlexS vil du ikke lige snakke med PO om det her? Hvis de er enige må du godt oprette issuet med alt den relevante information.

**Comment by ggAlexS** [comment]
> > > Desuden har jeg også observeret at når man tilføjer/slettet/kopierer aktiviteter i weekplanen, bliver alle choiceboard navne reset til "Choiceboard". Dette skyldes navnet ikke bliver taget med under aktiviteter kopieres under UpdateWeek i api'et. Synderen er en constructor i GirafRest/Shared/SharedMethods.cs linje 102. Api delen af det burde nok have været mere grundigt testet, så var denne fejl blevet fanget.
> 
> > 
> 
> > der bør nok laves et nyt issue på web-api'en til det her
> 
> 
> 
> @ggAlexS vil du ikke lige snakke med PO om det her? Hvis de er enige må du godt oprette issuet med alt den relevante information.

Altså ligemeget hvad skal backend problemer ikke løses på en weekplaner pull request. client-api og web-api delene af det her issues er kørt igennem for længe siden.

**Comment by ggAlexS** [comment]
> > > > Desuden har jeg også observeret at når man tilføjer/slettet/kopierer aktiviteter i weekplanen, bliver alle choiceboard navne reset til "Choiceboard". Dette skyldes navnet ikke bliver taget med under aktiviteter kopieres under UpdateWeek i api'et. Synderen er en constructor i GirafRest/Shared/SharedMethods.cs linje 102. Api delen af det burde nok have været mere grundigt testet, så var denne fejl blevet fanget.
> > 
> > 
> > > 
> > 
> > 
> > > der bør nok laves et nyt issue på web-api'en til det her
> > 
> > 
> > @ggAlexS vil du ikke lige snakke med PO om det her? Hvis de er enige må du godt oprette issuet med alt den relevante information.
> 
> Altså ligemeget hvad skal backend problemer ikke løses på en weekplaner pull request. client-api og web-api delene af det her issues er kørt igennem for længe siden.

Der er en pull request på web-api nu

**Comment by AdrianPlesner** [comment]
> Der er en pull request på web-api nu

@ggAlexS Den kan jeg da ikke finde?

**Comment by ggAlexS** [comment]
> 
> 
> > Der er en pull request på web-api nu
> 
> @ggAlexS Den kan jeg da ikke finde?

https://github.com/aau-giraf/web-api/pull/179

**Comment by AdrianPlesner** [comment]
> > > Der er en pull request på web-api nu
> > 
> > 
> > @ggAlexS Den kan jeg da ikke finde?
> 
> [aau-giraf/web-api#179](https://github.com/aau-giraf/web-api/pull/179)

👍

File: lib/screens/show_activity_screen.dart [hunk 3]
```diff
@@ -277,61 +272,96 @@ class ShowActivityScreen extends StatelessWidget {
...
-              );
-            });
+                );
+              });
         });
   }
 
   /// Builds the activity widget.
   Card buildActivity(BuildContext context) {
+    String inputtext = '';
```
Comment (AdrianPlesner) [inline lib/screens/show_activity_screen.dart]: Denne bliver reset til '' ved fx tryk på enter, eller hvis man kollapser sit onscreen keyboard

File: lib/screens/show_activity_screen.dart [hunk 4]
```diff
@@ -277,61 +272,96 @@ class ShowActivityScreen extends StatelessWidget {
...
+      Visibility(
+        visible: _activity.isChoiceBoard,
+        child: Row(
+          children: <Widget>[
+            Flexible(
+              child: TextFormField(
+                key: const Key('ChoiceBoardNameText'),
+                initialValue: _activity.choiceBoardName,
+                textAlign: TextAlign.center,
+                onChanged: (String text) {
```
Comment (AdrianPlesner) [inline lib/screens/show_activity_screen.dart:349]: Enten lad være med at bruge et FormField og indsæt en controller til at tilgå teksten ordentligt, eller rent faktisk brug Form funktionaliteten til noget e.g. wrap den i en Form hvorpå man kan kalde .save med gemknappen og bruge tekstfeltets onSave.

**Review by AdrianPlesner — CHANGES_REQUESTED** [review]
Der er nogle problemer når man trykker enter eller lukker sit keyboard, og evt andre ting der for skærmen til at refresh

**Comment by AdrianPlesner** [comment]
environments.json er vidst kommet med.

File: lib/screens/show_activity_screen.dart (hunk 3 above)
Comment (JeppeKLau) [reply lib/screens/show_activity_screen.dart]: Har fixet problemet med enter knappen, men det med at collapse keyboarded er by design. Hvis du lukker dit keyboard skal titlen på choice board ikke ændres.

File: lib/screens/show_activity_screen.dart (hunk 4 above)
Comment (JeppeKLau) [reply lib/screens/show_activity_screen.dart:349]: Man kan sagtens nøjes med den tilgang vi har valgt, da Flutter automatisk tilføjer en controller i tilfælde hvor et `TextFormField` bliver lavet uden at give en controller, som det fremgår af dokumentationen: `If [controller] is null, then a [TextEditingController] will be constructed automatically and its text will be initialized to [initialValue] or the empty string.`

File: lib/screens/show_activity_screen.dart (hunk 3 above)
Comment (AdrianPlesner) [reply lib/screens/show_activity_screen.dart]: Der er også fint, men hvis jeg trykker på Godkend knappen efter at jeg har lukket mit keyboard bliver titlen sat til `''` hvilket jeg ikke tænker var hensigten? Desuden giver det fejl så aktiviteten bliver rød når man vender tilbage til uge oversigten.

File: lib/screens/show_activity_screen.dart (hunk 4 above)
Comment (AdrianPlesner) [reply lib/screens/show_activity_screen.dart:349]: Det er også helt fint, jeg tænker bare ikke at der er noget grund til at bruge et TextFormField i stedet for et TextField, hvis man alligevel ikke bruger Form delen af den?

Comment (JeppeKLau) [reply lib/screens/show_activity_screen.dart:349]: Det er nemmest at bruge et `TextFormField` i forhold til at initialisere værdien af tekst feltet. Så selvom at al form funktionaliteten ikke bliver brugt, er det den nemmeste tilgang til at tilføje den her funktionalitet.

File: lib/screens/show_activity_screen.dart (hunk 3 above)
Comment (JeppeKLau) [reply lib/screens/show_activity_screen.dart]: Det burde være fikset nu

**Review by AdrianPlesner — APPROVED** [review]

**Comment by social2dk** [comment]
Jeg kan ikke få lov til at rename hvis jeg lige har tilføjet en valgmulighed, både hvis det er en som allerede har valgmuligheder, og en som ikke havde i forvejen. Er det et problem?
Altså hvis jeg ligesom går ud fra aktiviteten, så jeg er ude og se alle ugeplaner igen, og så går ind på choiceboardet igen, så kan jeg godt

**Comment by JeppeKLau** [comment]
> 
> 
> Jeg kan ikke få lov til at rename hvis jeg lige har tilføjet en valgmulighed, både hvis det er en som allerede har valgmuligheder, og en som ikke havde i forvejen. Er det et problem?
> Altså hvis jeg ligesom går ud fra aktiviteten, så jeg er ude og se alle ugeplaner igen, og så går ind på choiceboardet igen, så kan jeg godt

Lige umiddelbart, så virker det til at være et problem med måden at aktiviteter bliver opdateret når de bliver til choice boards eller får tilføjet nye valgmuligheder. Det er ikke noget der er direkte relateret til det vi har tilføjet. Overvej at oprette et issue på det, kunne forestille mig at det kan hjælpe hvis man implementerede en state og brugte den til at holde styr på aktiviteterne.

File: pubspec.lock [hunk 5]
```diff
@@ -30,7 +30,7 @@ packages:
       name: async
```
Comment (social2dk) [inline pubspec.lock:30]: Den her skal ikke med, skal den?

Comment (social2dk) [reply pubspec.lock:30]: Altså hele pubspec.lock

Comment (JeppeKLau) [reply pubspec.lock:30]: Jo den skal med, for den bliver opdateret automatisk af `flutter pub get` og den beskriver bare hvilken specifikke version af packages som er blevet hentet da `flutter pub get` er blevet kørt.

**Review by social2dk — APPROVED** [review]

**Pull request merged**